internal/dataplane/parser/translate_utils_test.go

```diff
@@ -170,3 +172,256 @@ func Test_convertGatewayMatchHeadersToKongRouteMatchHeaders(t *testing.T) {
+					{
+						Group: gatewayv1alpha2.Group(""),
+						Kind:  gatewayv1alpha2.Kind("PolicyTwo"),
+					},
+				},
+			},
+		},
+	}
+	tests := []struct {
+		msg    string
+		from   gatewayv1alpha2.ReferencePolicyFrom
+		result map[gatewayv1alpha2.Namespace][]gatewayv1alpha2.ReferencePolicyTo
+	}{
+		{
+			msg: "no matches whatsoever",
+			from: gatewayv1alpha2.ReferencePolicyFrom{
+				Group:     gatewayv1alpha2.Group("invalid.example"),
+				Kind:      gatewayv1alpha2.Kind("invalid"),
+				Namespace: gatewayv1alpha2.Namespace("invalid"),
+			},
+			result: map[gatewayv1alpha2.Namespace][]gatewayv1alpha2.ReferencePolicyTo{},
+		},
+		{
+			msg: "non-matching namespace",
+			from: gatewayv1alpha2.ReferencePolicyFrom{
+				Group:     gatewayv1alpha2.Group("gateway.networking.k8s.io"),
+				Kind:      gatewayv1alpha2.Kind("UDPRoute"),
+				Namespace: gatewayv1alpha2.Namespace("niyazi"),
+			},
+			result: map[gatewayv1alpha2.Namespace][]gatewayv1alpha2.ReferencePolicyTo{},
+		},
+		{
+			msg: "non-matching kind",
+			from: gatewayv1alpha2.ReferencePolicyFrom{
+				Group:     gatewayv1alpha2.Group("gateway.networking.k8s.io"),
+				Kind:      gatewayv1alpha2.Kind("TLSRoute"),
+				Namespace: gatewayv1alpha2.Namespace("behbudiy"),
+			},
+			result: map[gatewayv1alpha2.Namespace][]gatewayv1alpha2.ReferencePolicyTo{},
+		},
+		{
+			msg: "non-matching group",
+			from: gatewayv1alpha2.ReferencePolicyFrom{
+				Group:     gatewayv1alpha2.Group("invalid.example"),
+				Kind:      gatewayv1alpha2.Kind("UDPRoute"),
+				Namespace: gatewayv1alpha2.Namespace("behbudiy"),
+			},
+			result: map[gatewayv1alpha2.Namespace][]gatewayv1alpha2.ReferencePolicyTo{},
+		},
+		{
+			msg: "single match",
+			from: gatewayv1alpha2.ReferencePolicyFrom{
+				Group:     gatewayv1alpha2.Group("gateway.networking.k8s.io"),
+				Kind:      gatewayv1alpha2.Kind("UDPRoute"),
+				Namespace: gatewayv1alpha2.Namespace("behbudiy"),
+			},
+			result: map[gatewayv1alpha2.Namespace][]gatewayv1alpha2.ReferencePolicyTo{
+				"cholpon": {
+					{
+						Group: gatewayv1alpha2.Group(""),
+						Kind:  gatewayv1alpha2.Kind("PolicyTwo"),
+					},
+				},
+			},
+		},
+		{
+			msg: "multiple matches",
+			from: gatewayv1alpha2.ReferencePolicyFrom{
+				Group:     gatewayv1alpha2.Group("gateway.networking.k8s.io"),
+				Kind:      gatewayv1alpha2.Kind("TCPRoute"),
+				Namespace: gatewayv1alpha2.Namespace("qodiriy"),
+			},
+			result: map[gatewayv1alpha2.Namespace][]gatewayv1alpha2.ReferencePolicyTo{
+				"cholpon": {
+					{
+						Group: gatewayv1alpha2.Group(""),
+						Kind:  gatewayv1alpha2.Kind("PolicyTwo"),
+					},
+				},
+				"fitrat": {
+					{
+						Group: gatewayv1alpha2.Group(""),
+						Kind:  gatewayv1alpha2.Kind("PolicyOne"),
+					},
+				},
+			},
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.msg, func(t *testing.T) {
+			result := getReferenceTosForFrom(tt.from, policies)
+			assert.Equal(t, tt.result, result)
+		})
+	}
+}
```

Comment:
Unless I've missed it somewhere, we're missing unit tests for the new `generateKongServiceFromBackendRef` method.

Reply:
Do we think we need one? There were no units for its predecessors. It requires a lot of inputs and a unit wouldn't really tell us much anything different than the integration test ("we got this set of services" versus "we know we go this set of services because we do/do not route to them"), and that saves us having to build the intermediate parser inputs/store/etc.

Reply:
I personally think it would be good to add unit tests for anything we can unit test unless we have a strong reason not to.

Reply:
Can now confirm that constructing these inputs from scratch rather than using proper objects retrieved from the API client is :grimacing: